feat: locally cache frequently requested LDAP mapping data

blizzz opened the pull request:
<!--
  - 🚨 SECURITY INFO
  -
  - Before sending a pull request that fixes a security issue please report it via our HackerOne page (https://hackerone.com/nextcloud) following our security policy (https://nextcloud.com/security/). This allows us to coordinate the fix and release without potentially exposing all Nextcloud servers and users in the meantime.
-->

## Summary

We save that looking up DN by user/group ID is taking up to 5% of all database queries. Albeit they are simple and  do not transfer much data, it is still worthwhile to reduce them, since this data barely ever changes.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [X] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

blizzz committed:
feat: locally cache frequently requested LDAP mapping data

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

come-nc reviewed (commented):
I tend to be a bit scared about this kind of things, we cache stuff from LDAP to DB to avoid too many LDAP queries, and then we cache stuff from DB to Cache to avoid DB queries.

Why local cache and not distributed? Is it because the cache is local that there is the FIXME about broadcasting?

Why the limit to 2000 by default, what is the risk of putting too much lines in Cache? 1MiB does not sound like that much.

blizzz commented on apps/user_ldap/lib/Mapping/AbstractMapping.php:
```diff
@@ -27,16 +31,76 @@ abstract class AbstractMapping {
-	/** @var array caches Names (value) by DN (key) */
-	protected $cache = [];
+	protected function initLocalCache(): void {
+		if ($this->isCLI || !$this->cacheFactory->isLocalCacheAvailable()) {
```
For APCu does not have a shared cache. There is no benefit over the runtime cache that was already present.

blizzz commented on apps/user_ldap/lib/Mapping/AbstractMapping.php:
```diff
@@ -27,16 +31,76 @@ abstract class AbstractMapping {
+			return;
+		}
+
+		$section = \str_contains($this->getTableName(), 'user') ? 'u/' : 'g/';
```
> Why make it more complicated than this?

It is shorter and thus may save allocation of bytes in APCU.

blizzz commented on apps/user_ldap/lib/Mapping/AbstractMapping.php:
```diff
@@ -27,16 +31,76 @@ abstract class AbstractMapping {
+		$result->closeCursor();
+
+		$use = (int)$row['count'] <= self::LOCAL_CACHE_OBJECT_THRESHOLD;
+		$this->localNameToDnCache->set('use', $use, intdiv(self::LOCAL_CACHE_TTL, 4));
```
User count is more likely to fluctuate, so this is a lazy attempt to make it a weekly check. Should result in a different constant.

blizzz committed:
refactor(LDAP): switch from prepares statement to query builder

- has the advantage that queries will be reported in the query.log when
  configured

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

artonge reviewed (approved):
Looks good to me besides nitpick

Does it mean that admin would need to flush the cache if they change one user's display name? If so, we'll need to update the documentation.

blizzz commented:
I think I can simulate a case where 
- the local cache is initilized with false
- the dn has changed

that fails. It is a bit constructed. And actually, I can reproduce with plain master.

But I don't like the way it fails and getting out of this scenario might not be inconvenient.

blizzz commented:
> I think I can simulate a case where
> 
>     * the local cache is initilized with false
> 
>     * the dn has changed
> 
> 
> that fails. It is a bit constructed. And actually, I can reproduce with plain master.
> 
> But I don't like the way it fails and getting out of this scenario might not be inconvenient.

Solved with https://github.com/nextcloud/server/pull/54429/files#diff-607694866b0eede88107508c374cd8cb64bd1518dda2bd98a6dae3dae25c4e27R73-R75
